proposals/p1382.md

```diff
@@ -0,0 +1,108 @@
+| 2010       | Rust       | explicit                             | `self`         | `Self`        |
+| 2011       | Kotlin     | implicit                             | `this`         | ---           |
+| 2012       | TypeScript | explicit                             | `this`         | `this`        |
+| 2014       | Swift      | implicit                             | `self`         | `Self`        |
+| previously | Carbon     | explicit                             | `me`           | `Self`        |
+| proposed   | Carbon     | explicit                             | `self`         | `Self`        |
+
+In detail:
+
+-   C++ uses
+    [`this` for address of the receiver value](https://en.cppreference.com/w/cpp/language/this)
+-   Swift uses
+    [`self` for the receiver value](https://docs.swift.org/swift-book/LanguageGuide/Methods.html#ID238).
+    and
+    [`Self` for its type](https://docs.swift.org/swift-book/ReferenceManual/Types.html#ID610).
+-   Rust uses
+    [`self` for the receiver value](https://doc.rust-lang.org/std/keyword.self.html)
+    and
+    [`Self` for its type](https://doc.rust-lang.org/rust-by-example/fn/methods.html).
+-   C# uses
+    [`this` for a reference to the receiver value](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/this)
+-   Python as a convention uses `self` for the receiver value, but it's almost
+    universally followed.
+-   Go conventionally uses an abbreviation of the type name.
+
+## Proposal
+
+Use `self` instead of `me` to be consistent with Swift and Rust.
+
+## Rationale
+
+This is consistent with Carbon's goal to make
+[Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+by choosing a keyword for this role that is less surprising to users.
+
+## Alternatives considered
+
+### Don't change anything
+
+We could stay with the status quo, which has the benefit that `me` is shorter
+than `self`. There are two considerations:
+
+-   For accessing members of the current object, the chart in
+    [the background section](#background) shows plenty of precedent for
+    requiring a 4 character explicit keyword.
+-   We would also like to reduce ceremony when declaring the signature of a
+    method. For this concern, both `me: Self` and `addr me: Self*` are already
+    longer than what other languages use in practice. It would probably be
+    better to solve this problem with a shortcut approach like Rust
+    ([1](https://doc.rust-lang.org/book/ch05-03-method-syntax.html),
+    [2](https://doc.rust-lang.org/rust-by-example/fn/methods.html)), where
+    `&self` is short for `self: &Self` and `&mut self` is short for
+    `self: &mut Self`.
+
+### `this`
+
+We could also switch to `this`, primarily to benefit
+[C++ users](https://en.cppreference.com/w/cpp/language/this). However, we are
+worried that it frequently not being a pointer would be surprising to those C++
+users.
```

Comment:
I think this rationale should be expanded because of the break in consistency with C++. Yes, it won't always be the same type: but most C++ developers are familiar with the basic meaning of `this`, and the type would be in the function signature.

For example, although you don't note the rationale here, I think you'd noted that `self` operates this way in Rust. However, that's Rust, and most C++ developers shouldn't be expected to be familiar with it. The surrounding syntax also differs from Rust, which I would expect to hamper experienced Rust devs.

As a counterpoint to the Rust type argument, more likely C++ developers would be familiar with `self` in Python, due to its popularity, and it's always the same type there. From what I can tell, Swift is similar. Thus I don't see `self` as offering better cross-language consistency from a type perspective, since Rust is in a minority.

I believe more information should be provided for the leads here -- it may even be appropriate to raise the question for explicit resolution, although it's perhaps redundant with this proposal.

Reply:
In the 'cons' colums, it feels less clear to, from a semantic standpoint, that `this` refers to the current instance. `self` feels like it less ambiguously refer 'itself' vs designating somethig with 'this'/'that'.

Reply:
> we are worried that it frequently not being a pointer would be surprising to those C++ users

Could expand this alternative to also suggest relying on user feedback on whether `this` not being a pointer would be surprising, rather than assumption.

Reply:
I've expanded in this update: https://github.com/carbon-language/carbon-lang/pull/1382/commits/fda1ceb11525c2c4ff0e1d3ce1c0e7d7e610b316

Reply:
Languages with self-referential feature using keywords:
- self: Modula-3, Objective-C, Python, Ruby, Smalltalk
- this: C#, C++, Java 

Feel free to fix or append to these lists above
[Very short history of "self"](https://stackoverflow.com/a/1080192/624900)

Reply:
I've added that link to history in the background section: see https://github.com/carbon-language/carbon-lang/pull/1382/commits/2b7ee5ad82e7d9ca9052110939c6be64367c451d